src/CrosschainDeployScript.sol

```diff
@@ -0,0 +1,116 @@
+     * @param contractString Contract name in the form of `ContractFile.sol`, if the name of the contract and the file are the same, or `ContractFile.sol:ContractName` if they are different.
+     * @param gasLimit Contract deploy and init gas.
+     * @param salt Entropy for contract address generation.
+     * @param isUniquePerChain True to have unique addresses on every chain.
+     * @param constructorArgs Bytes to add to the deployBytecode, or empty, one per chain.
+     * @param initDatas Bytes to send to the contract after deployment, or empty, one per chain.
+     *
+     *   Users call this function and pass only the function call string as
+     *   `MyContract.sol:MyContract`. The function call string is then parsed
+     *   and the `callData` and `bytesCode` are extracted from it.
+     *   and the contract is deployed on the other chains.
+     */
+    function deploy(
+        string calldata contractString,
+        uint256 gasLimit,
+        bytes32 salt,
+        bool isUniquePerChain,
+        bytes[] memory constructorArgs,
+        bytes[] memory initDatas
+    )
+        // uint8[] memory destinationDomainIDs
+        public
+        payable
+    {
+        require(_deploymentNetworks.length > 0, "Need to add deployment targets. Use `addDeploymentTarget` first");
+        // We use the contractString to get the bytecode of the contract,
+        // reference: https://book.getfoundry.sh/cheatcodes/get-code
+        bytes memory deployByteCode = vm.getCode(contractString);
+        uint256[] memory fees = ICrosschainDeployAdapter(CROSS_CHAIN_DEPLOY_CONTRACT_ADDRESS).calculateDeployFee(
+            deployByteCode, gasLimit, salt, isUniquePerChain, constructorArgs, initDatas, _deploymentNetworks
+        );
+        uint256 totalFee;
+        uint256 feesArrayLength = fees.length;
+        for (uint256 i = 0; i < feesArrayLength;) {
+            uint256 fee = fees[i];
+            totalFee += fee;
+            unchecked {
+                ++i;
+            }
+        }
+
+        ICrosschainDeployAdapter(CROSS_CHAIN_DEPLOY_CONTRACT_ADDRESS).deploy{value: totalFee}(
+            deployByteCode, gasLimit, salt, isUniquePerChain, constructorArgs, initDatas, _deploymentNetworks, fees
+        );
+    }
+
+    /**
+     * This function willl take the network, constructor args and initdata and
+     * save these to a mapping (what type?)
+     */
+    function addDeploymentTarget(string memory deploymentTarget) public {
+        uint8 deploymentTargetDomainId = _stringToDeploymentNetwork[deploymentTarget];
+        require(deploymentTargetDomainId != 0, "Invalid deployment target");
+        if ((deploymentTargetDomainId == 3) || (deploymentTargetDomainId == 4)) {
+            revert Unimplemented("That domain isn't implemented yet");
+        }
+
+        _deploymentNetworks.push(deploymentTargetDomainId);
+        // TODO: what else does this need to do?
+    }
+
+    // what does this do?
+    function encodeInitData() public {}
```

Comment:
ideally takes name of the contract method and args and encodes function call to that method. On deploying this method call would be invoked after contract is deployed

Reply:
Foundry does not seem to provide access to contract interface by string names. For Solidity developers it would be easier to just do:
```
abi.encodeCall(MyContract.myFunc, arg1, arg2, arg3, ...);
```

Reply:
@lastperson shame. Oh well, @stonecharioteer just take encoded call as param but make sure to add @lastperson example in the docs^^